python/perforce.py

```diff
@@ -355,26 +355,29 @@ def p4print_unshelve(self, changelist):
 
         if 'depotFile' not in changeinfo:
             raise Exception('Changelist %s does not contain any shelved files' % changelist)
+
+        sync_prefixes = [prefix.rstrip('.') for prefix in self.sync_paths]
+        self.perforce.logger.info("Filtering changelist against the "
+                                  "following sync paths: %s", sync_prefixes)
+
         depotfiles = changeinfo['depotFile']
 
         # Filter out files with exclusive locking
         filtered_files = []
         exclusive_files = []
         for i in range(len(depotfiles)):
-            if "+l" in changeinfo["type"][i]:
-                exclusive_files.append(depotfiles[i])
-                continue
-            filtered_files.append(depotfiles[i])
+            if (any(depotfiles[i].startswith(prefix)
+                    for prefix in sync_prefixes)):
```

Comment:
Unfortunately, due to the structure of the data we get from p4plugin, we have to map files to type strings by way of index. Which means we have to filter against `sync_prefixes` within the same loop.